### PR TITLE
update run implementation

### PR DESCRIPTION
**Pull request by ryanzhang-oss**
### Description of your changes

Fixes #

I have:

- [ ] Run `make reviewable` to ensure this PR is ready for review.

### CODE REVIEW

File: pkg/controllers/updaterun/executing.go [hunk 1]
```diff
@@ -0,0 +1,479 @@
...
+			// delete the binding that is up to date with the cluster status.
+			if err := r.Client.Delete(ctx, binding); err != nil {
+				klog.ErrorS(err, "Failed to delete a binding in the update run", "clusterResourceBinding", klog.KObj(binding), "cluster", curCluster.ClusterName, "stagedUpdateRun", klog.KObj(updateRun))
+				return false, controller.NewAPIServerError(false, err)
+			}
+			klog.V(2).InfoS("Deleted a binding pointing to a to be deleted cluster", "clusterResourceBinding", klog.KObj(binding), "cluster", curCluster.ClusterName, "stagedUpdateRun", klog.KObj(updateRun))
+			markClusterUpdatingStarted(curCluster, updateRun.Generation)
+			if deletingBinding == 0 {
+				markStageUpdatingStarted(updateRun.Status.DeletionStageStatus, updateRun.Generation)
+			}
```
Comment (jwtty) [inline pkg/controllers/updaterun/executing.go:307]: We should `deletingBinding++` here.

File: pkg/controllers/updaterun/executing.go [hunk 2]
```diff
@@ -0,0 +1,479 @@
...
+			// the binding is not up to date with the cluster status, it could happen if there are multiple update stage running concurrently.
+			delErr := fmt.Errorf("the deleted cluster `%s` in the deleting stage does not match the cluster status: %+v, binding := %+v", curCluster.ClusterName, curCluster, binding.Spec)
+			klog.ErrorS(delErr, "The binding has been changed after the updating, please check if there is con-current update run", "stagedUpdateRun", klog.KObj(updateRun))
+			return false, fmt.Errorf("%w: %s", errStagedUpdatedAborted, delErr.Error())
+		}
+	}
+	// the rest of the clusters in the stage are not in the tobeDeletedBindings so it should be marked as delete succeeded
+	for _, clusterStatus := range existingDeleteStageClustersMap {
+		// make sure the cluster is marked as deleting
+		if !condition.IsConditionStatusTrue(meta.FindStatusCondition(clusterStatus.Conditions, string(placementv1alpha1.ClusterUpdatingConditionStarted)), updateRun.Generation) {
```
Comment (jwtty) [inline pkg/controllers/updaterun/executing.go:318]: Is it possible for this condition to happen? The ClusterUpdating never started?

File: apis/placement/v1alpha1/stagedupdate_types.go [hunk 3]
```diff
@@ -18,21 +18,22 @@ import (
...
+// selected resources to specified clusters.
 // Resources from unselected clusters are removed after all stages in the update strategy are completed.
-// Each StagedUpdateRun object corresponds to a single release of a specific resource version.
-// The release is abandoned if the StagedUpdateRun object is deleted or the scheduling decision changes.
-// The name of the StagedUpdateRun must conform to RFC 1123.
-type StagedUpdateRun struct {
+// Each ClusterResourceStagedUpdateRun object corresponds to a single release of a specific resource version.
+// The release is abandoned if the ClusterResourceStagedUpdateRun object is deleted or the scheduling decision changes.
+// The name of the ClusterResourceStagedUpdateRun must conform to RFC 1123.
+type ClusterResourceStagedUpdateRun struct {
```
[the remark this answers is not in the document]
Comment (jwtty) [reply apis/placement/v1alpha1/stagedupdate_types.go]: Outdated.

File: apis/placement/v1alpha1/stagedupdate_types.go [hunk 4]
```diff
@@ -157,19 +158,25 @@ type StagedUpdateRunStatus struct {
 	// +kubebuilder:validation:Optional
 	PolicySnapshotIndexUsed string `json:"policySnapshotIndexUsed,omitempty"`
 
+	// PolicyObservedCRPGeneration is the generation of the CRP which the scheduler uses to perform the scheduling.
+	// It is computed at the beginning of the update run from the policy snapshot object. If the CRP is updated
+	// during the update run in a way that causes the policy snapshot condition to change, the update run is abandoned.
+	// +kubebuilder:validation:Optional
+	PolicyObservedNodeCount int `json:"policyObservedNodeCount,omitempty"`
```
[the remark this answers is not in the document]
Comment (jwtty) [reply apis/placement/v1alpha1/stagedupdate_types.go]: Outdated.

File: pkg/controllers/updaterun/executing.go [hunk 5]
```diff
@@ -0,0 +1,479 @@
...
+func (r *Reconciler) checkAfterStageTasksStatus(ctx context.Context, updatingStageIndex int, updateRun *placementv1alpha1.ClusterStagedUpdateRun) (bool, error) {
+	updatingStageStatus := &updateRun.Status.StagesStatus[updatingStageIndex]
+	updatingStage := &updateRun.Status.StagedUpdateStrategySnapshot.Stages[updatingStageIndex]
+	if updatingStage.AfterStageTasks == nil {
+		klog.V(2).InfoS("There is no after stage task for this stage", "stage", updatingStage.Name, "stagedUpdateRun", klog.KObj(updateRun))
+		return true, nil
+	}
+	for i, task := range updatingStage.AfterStageTasks {
+		switch task.Type {
+		case placementv1alpha1.AfterStageTaskTypeTimedWait:
```
[the remark this answers is not in the document]
Comment (ryanzhang-oss) [reply pkg/controllers/updaterun/executing.go:198]: need to find a way to ban that

**Pull request closed without merging**